### ISSUE TITLE
Double compilation during bootstrap

### ISSUE THREAD

**Issue by nasser** [nasser/magic#207]
The following namespaces are being double compiled during bootstrapping. 

* `clojure.clr.io`
* `clojure.core.protocols`
* `clojure.edn`
* `clojure.instant`
* `clojure.main`
* `clojure.set`
* `clojure.string`
* `clojure.uuid`
* `clojure.walk`

**Comment by skydread1** [comment]
## Explanation

The double compilation occurs when
- the ns was required by another ns (and so got compiled automatically)
- `(compile 'required-ns)` was called after `(compile 'requiring-ns)`

## Example

`clojure.core.protocols` requires `clojure.set`

so if the compilation order is
```clojure
(compile 'clojure.core.protocols)
(compile 'clojure.set)
```
`clojure.set` will be compiled twice (once because required in `clojure.core.protocols` and once on its own.

but if the compilation order is

```clojure
(compile 'clojure.set)
(compile 'clojure.core.protocols)
```
`clojure.set` will be compiled once.

That's my understanding.

**Comment by nasser** [comment]
if i recall the expected semantics, `compile` should see that `clojure.set` was already compiled and not compile it a second time. does it double compile on the JVM?

**Comment by nasser** [comment]
Fixed by 8d52baab28c7b78c2d16dc4418bae1c59059edb8
